### PR TITLE
Doc the fields of some libgit2 types

### PR DESCRIPTION
**Pull request by kshyatt**
cf #18810

### CODE REVIEW

File: base/libgit2/types.jl [hunk 1]
```diff
@@ -287,6 +299,12 @@ end
     LibGit2.DescribeFormatOptions
 
 Matches the [`git_describe_format_options`](https://libgit2.github.com/libgit2/#HEAD/type/git_describe_format_options) struct.
+
+The fields represent:
+  * `version`: version of the struct in use, in case this changes later. For now, always `1`.
+  * `abbreviated_size`: lower bound on the size of the abbreviated `GitHash` to use, defaulting to `7`.
+  * `always_use_long_format`: set to `1` to use the long format for strings even if a short format can be used.
+  * `dirty_suffix`: if set, if the [`workdir`](@ref) is dirty this will be appended to the description string.
```
Comment (tkelman) [inline base/libgit2/types.jl]: reorder this? if set, this will be appended ... if the workdir is dirty

File: base/libgit2/types.jl [hunk 2]
```diff
@@ -300,6 +318,17 @@ end
...
 Description of one side of a delta.
 Matches the [`git_diff_file`](https://libgit2.github.com/libgit2/#HEAD/type/git_diff_file) struct.
+
+The fields represent:
+  * `id`: the [`GitHash`](@ref) of the item in the diff. If the item is empty on this
+     side of the diff (for instance, if the diff is of the removal of a file), this will
+     be `GitHash(0)`.
+  * `path`: a `NULL` terminated path to the item relative to the working directory of the repository.
+  * `size`: the size of the item in bytes.
+  * `flags`: a combination of the [`git_diff_flag_t`](https://libgit2.github.com/libgit2/#HEAD/type/git_diff_flag_t) flags.
```
Comment (tkelman) [inline base/libgit2/types.jl]: combined how?

File: base/libgit2/types.jl [hunk 3]
```diff
@@ -300,6 +318,17 @@ end
...
+  * `mode`: the [`stat`](@ref) mode for the item.
+  * `id_abbrev`: only present in versions newer than or equal to `0.25.0`.
```
Comment (tkelman) [inline base/libgit2/types.jl]: libgit2 versions

**Comment by kshyatt** [comment]
Got everything?

File: base/libgit2/types.jl [hunk 4]
```diff
@@ -193,6 +193,18 @@ end
 Options for connecting through a proxy.
 
 Matches the [`git_proxy_options`](https://libgit2.github.com/libgit2/#HEAD/type/git_proxy_options) struct.
+
+The fields represent:
+  * `version`: version of the struct in use, in case this changes later. For now, always `1`.
+  * `proxytype`: the type of proxy to use. Default is to auto-detect it.
```
Comment (tkelman) [inline base/libgit2/types.jl]: this could maybe use a xref to the enum that the values should come from

**Comment by kshyatt** [comment]
Again checked `make check-whitespace` locally.

File: base/libgit2/types.jl (hunk 4 above)
Comment (tkelman) [inline base/libgit2/types.jl]: there's also a Julia enum, right?

**Review by ararslan — APPROVED** [review]

File: base/libgit2/types.jl [hunk 5]
```diff
@@ -193,6 +193,20 @@ end
 Options for connecting through a proxy.
 
 Matches the [`git_proxy_options`](https://libgit2.github.com/libgit2/#HEAD/type/git_proxy_options) struct.
+
+The fields represent:
+  * `version`: version of the struct in use, in case this changes later. For now, always `1`.
+  * `proxytype`: an `enum` for the type of proxy to use.
+     Defined in [`git_proxy_t`](https://libgit2.github.com/libgit2/#HEAD/type/git_proxy_t).
```
Comment (tkelman) [inline base/libgit2/types.jl]: does the Julia enum use the same spellings? worth spelling out the qualified Julia-side enum name those would be specified as

Comment (kshyatt) [inline base/libgit2/types.jl]: Done - look ok?

Comment (kshyatt) [inline base/libgit2/types.jl]: `make check-whitespace` passed again locally

Comment (tkelman) [inline base/libgit2/types.jl]: looks great, though if it's not exported, maybe qualify it with the module name?

Comment (kshyatt) [inline base/libgit2/types.jl]: If you're willing to wait, I hope to do another PR later tonight with examples that would show that it's not exported?

Comment (tkelman) [inline base/libgit2/types.jl]: ok

Comment (kshyatt) [inline base/libgit2/types.jl]: Example added!

File: base/libgit2/types.jl [hunk 6]
```diff
@@ -300,6 +333,18 @@ end
...
+  * `id`: the [`GitHash`](@ref) of the item in the diff. If the item is empty on this
+     side of the diff (for instance, if the diff is of the removal of a file), this will
+     be `GitHash(0)`.
+  * `path`: a `NULL` terminated path to the item relative to the working directory of the repository.
+  * `size`: the size of the item in bytes.
+  * `flags`: a combination of the [`git_diff_flag_t`](https://libgit2.github.com/libgit2/#HEAD/type/git_diff_flag_t)
+     flags. The `i`th bit of this integer sets the `i`th flag.
+  * `mode`: the [`stat`](@ref) mode for the item.
+  * `id_abbrev`: only present in LibGit2 versions newer than or equal to `0.25.0`.
+     The length of the `id` field when converted using [`hex`](@ref). Usually equal to `GIT_OID_HEXSZ`.
```
Comment (tkelman) [inline base/libgit2/types.jl]: is GIT_OID_HEXSZ a C name or the Julia enum name?

**Comment by tkelman** [comment]
doc build is failing here, one of the ref's might not be getting resolved?

**Comment by kshyatt** [comment]
Yeah, it was the link to `GitProxy`. I removed it because we have a description of all the options inline anyways.

File: base/libgit2/types.jl (hunk 6 above)
Comment (kshyatt) [inline base/libgit2/types.jl]: For us it's `OID_HEXSZ`. I'll update.

File: base/libgit2/consts.jl [hunk 7]
```diff
@@ -351,6 +351,13 @@ These are used to select which global option to set or get and are used in `git_
                     SET_SSL_CERT_LOCATIONS   = 12)
 
 
+    """
+Option flags for `GitProxy`.
```
Comment (tkelman) [inline base/libgit2/consts.jl]: hm this file is indented kind of strangely

Comment (kshyatt) [inline base/libgit2/consts.jl]: It sure is

**Comment by kshyatt** [comment]
Anything else I should do?

File: base/libgit2/types.jl [hunk 8]
```diff
@@ -193,6 +193,33 @@ end
...
+  * `version`: version of the struct in use, in case this changes later. For now, always `1`.
+  * `proxytype`: an `enum` for the type of proxy to use.
+     Defined in [`git_proxy_t`](https://libgit2.github.com/libgit2/#HEAD/type/git_proxy_t).
+     The corresponding Julia enum is `GIT_PROXY` and has values:
+     - `PROXY_NONE`: do not attempt the connection through a proxy.
+     - `PROXY_AUTO`: attempt to figure out the proxy configuration from the git configuration.
+     - `PROXY_SPECIFIED`: connect using the URL given in the `url` field of this struct.
+     Default is to auto-detect the proxy type.
+  * `url`: the URL of the proxy.
+  * `credential_cb`: a pointer to a callback function which will be called if the remote
```
Comment (tkelman) [inline base/libgit2/types.jl]: on the julia side this is just a function input, not a pointer right?

File: base/libgit2/consts.jl (hunk 7 above)
Comment (tkelman) [inline base/libgit2/consts.jl]: could fix it if you or someone else is feeling charitable, doesn't have to be in this pr tho

File: base/libgit2/types.jl (hunk 8 above)
Comment (kshyatt) [inline base/libgit2/types.jl]: Nope, `credential_cb::Ptr{Void}`

Comment (tkelman) [inline base/libgit2/types.jl]: do you need to do something with `cfunction` then to pass a julia function to it?

Comment (kshyatt) [inline base/libgit2/types.jl]: Yeah I think you have to do something much like the stuff in `base/libgit2/callbacks.jl`

**Comment by kshyatt** [comment]
Anything else I should change?

**Review by Sacha0 — APPROVED** [review]
superficially lgtm! :)

**Pull request merged**